### PR TITLE
monasca-schema, Mysql and PostgresSQL changes

### PR DESCRIPTION
**Pull request by kornicameister**
- updated text columns' lengths
- switched from enum to triggers in mysql schema
- switched from mediumtext to longtext in mysql schema
- optional task migration via ansible
- default migration files for mysql due to schema changes
 - column types
 - triggers

Change-Id: Ib72a36492f555da6154de30c7f057dac238f82a1

### CODE REVIEW

File: templates/mon_mysql.sql.j2 [hunk 1]
```diff
@@ -6,7 +6,7 @@ SET foreign_key_checks = 0;
 CREATE TABLE `alarm` (
   `id` varchar(36) COLLATE utf8mb4_unicode_ci NOT NULL,
   `alarm_definition_id` varchar(36) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '',
-  `state` enum('UNDETERMINED','OK','ALARM') COLLATE utf8mb4_unicode_ci NOT NULL,
```
Comment (tkuhlman) [inline templates/mon_mysql.sql.j2]: Why did you remove the enum?

Comment (kornicameister) [inline templates/mon_mysql.sql.j2]: Hibernate does not offer portable solution to map such column in the model. I mean, there's a way to do that but it works with MySQL only. We are trying to offer as portable Hibernate's model as possible but it wouldn't be possible if enums would be kept here. 

For reasons I cannot understand MySQL engines do not support CHECK clause we tried to replace enum with thus triggers were only reliable alternative to force constraints you had.  

The advantage is that schema in the form we pushed with this pull-request has better portability.

Comment (tkuhlman) [inline templates/mon_mysql.sql.j2]: Great thanks for the explanation that sounds good to me. The entire change looks fine I just need to test it now.

Comment (kornicameister) [inline templates/mon_mysql.sql.j2]: Great ! Waiting for feedback.

Comment (rbrndt) [inline templates/mon_mysql.sql.j2]: I'm not convinced triggers are the best option here. I would rather see a new table for the values in the enum, and a foreign key constraint referencing that new table. This makes the connection explicit and easily visible to anyone using the database, as well as modifiable without a schema update.

Comment (kornicameister) [inline templates/mon_mysql.sql.j2]: That was our idea as well but we decided that it'd be better to add triggers instead of adding new tables. Basically because of easier migration. 

Anyway we will investigate that, first thing Monday morning. That seems to be possible to map such schema in Hibernate because an enum (considered as Java type) can be mapped in two ways. For now we have a string, but there's an option to use a number that would be understood in schema as foreign key.

Comment (kornicameister) [inline templates/mon_mysql.sql.j2]: @rbrndt, We've made some drafts, some investigations and basically it is achievable but effort we would have to put in to make it work is tremendously greater. Not only migration would be harder, more vulnerable but there would be changes in at least monsaca-api and monasca-thresh as there are repositories (DAOs) working with plain MySQL (notice change in columns from varchars/enums to integer). Not the mention that the same thing could happen for other projects using the same database.

@tkuhlman, Do you think that transition to ref tables is something monasca would want to have instead of varchar based columns and triggers to force constraints ?

Comment (tkuhlman) [inline templates/mon_mysql.sql.j2]: I haven't researched the issue at all and don't know what is best. @roland-hochmuth do you want to take a moment to look at this and give your thoughts?

**Comment by ddieterly** [comment]
I think we need to use reference tables instead of triggers to enforce integrity constraints. This is the usually preferred solution unless there is a compelling reason to do otherwise. In this case there does not seem to be a compelling reason to use triggers. Triggers have the following drawbacks:

1. They have to be hand-coded and maintained. This places the burden on the developers who are most likely going to get it wrong at some point doing development/maintenance.

2. They are more difficult to understand and it can become very difficult to follow the logic. They complicate things enormously if propagated to a certain extent.

3. They require developers to understand and use the 'trigger' language.

Reference tables on the other hand are easy to understand and maintain. They use the standard facilities of the database in a natural way. No coding is necessary to add values to the reference tables.

The existing monasca code should not need to be changed to use reference tables. The primary key values in the reference tables can be strings and a foreign key integrity constraint can be instantiated just using the values of the strings. No integer-to-string conversions should be necessary. This will be less efficient than using an integer as the primary key, but these tables are going to be so small, that it probably won't make any difference.

**Comment by kornicameister** [comment]
- code rebased to github/master
- added ref tables + migration script

**Comment by ddieterly** [comment]
The following two tables should be a single table.

alarm_action_state
alarm_state

We need to ensure that the values are always the same. Two independent tables would allow the values to diverge.

Please merge these table into a single table.

**Comment by ddieterly** [comment]
Please create a pull request for the hpcloud/monasca-ci repo. The new tables that have been added in MySQL need to be added in the  monasca-ci/tests/smoke/smoke2_configs.py.j2 file. Just add them to the mysql_schema array. This test is currently failing with this pull request.

**Comment by kornicameister** [comment]
@ddieterly, tables have been merged + migration script has been updated.

**Comment by kornicameister** [comment]
https://github.com/hpcloud-mon/monasca-ci/pull/5

added tables in monasca-ci

**Comment by kornicameister** [comment]
Nice :+1:
